### PR TITLE
Initial import of Anki flashcard app

### PR DESCRIPTION
**Pull request by dsd**
Anki is a popular app for remembering things through the use of flashcards. It would be great to have it available via flathub. Here is my attempt at packaging it.

cc original app author @dae

### CODE REVIEW

File: net.ankiweb.Anki.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
+{
+  "app-id": "net.ankiweb.Anki",
+  "runtime": "org.kde.Platform",
+  "runtime-version": "5.9",
+  "sdk": "org.kde.Sdk",
+  "command": "anki",
+  "finish-args": [
+     "--socket=x11",
+     "--share=network"
```
Comment (TingPing) [inline net.ankiweb.Anki.json]: You'll want `--socket=wayland`, `--share=ipc`, and `--device=dri` also.

**Comment by dsd** [comment]
Anki would normally install mime type associations via https://github.com/dae/anki/blob/master/anki.xml
```
	xdg-mime install anki.xml --novendor
	xdg-mime default anki.desktop application/x-anki
	xdg-mime default anki.desktop application/x-apkg
```
How can I integrate this correctly in the manifest? Currently I remove those lines from the Makefile.

Also the build process complains confusingly about the icon:
```
Renaming anki.desktop to net.ankiweb.Anki.desktop
Committing stage cleanup to cache
Finishing app
Exporting share/applications/net.ankiweb.Anki.desktop
Exporting share/icons/net.ankiweb.Anki.png
Please review the exported files and the metadata
Committing stage finish to cache
Exporting net.ankiweb.Anki to repo
WARNING: Icon referenced in desktop file but not exported: net.ankiweb.Anki
```

It complains that it was not exported 4 lines after having exported it?

File: net.ankiweb.Anki.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      ],
+      "buildsystem": "simple",
+      "build-commands": [
+        "mkdir -p /app/local/bin",
+        "mkdir -p /app/share/man/man1",
+        "mkdir -p /app/share/{applications,icons,pixmaps}",
+        "sed -i -e 's/Icon=anki/Icon=net.ankiweb.Anki/' anki.desktop",
+        "sed -i -e 's:/usr/share:/app/share:' tools/runanki.system",
+        "sed -i -e '/xdg-mime/d' Makefile",
+        "make install PREFIX=/app",
```
[the remark this answers is not in the document]
Comment (TingPing) [reply net.ankiweb.Anki.json]: You can patch it here first but it would be great if it made it upstream eventually.

**Comment by TingPing** [comment]
> Anki would normally install mime type associations via https://github.com/dae/anki/blob/master/anki.xml

Flatpak's can still export mime assocations the installed file just needs to be renamed to the app-id like everything else (make not of icon references in them also).

The `xdg-mime` tool not being there might be annoying, maybe just package it as a build dep, maybe just patch it to directly `install` the file (I don't think `xdg-mime install` does anything of value)

> It complains that it was not exported 4 lines after having exported it?

Because `Exporting share/icons/net.ankiweb.Anki.png` is the wrong path. `share/icons/hicolor/$ICON_SIZE/apps/net.ankiweb.Anki.png` is correct.

File: net.ankiweb.Anki.json [hunk 3]
```diff
@@ -64,7 +64,8 @@
       "buildsystem": "simple",
       "build-commands": [ 
         "python3 configure.py --assume-shared --confirm-license --no-designer-plugin --no-qml-plugin --sysroot=/app -w --sip=/app/bin/sip --sip-incdir=/app/include -verbose QMAKE_CFLAGS_RELEASE='-I/usr/include/python3.5m/' QMAKE_CXXFLAGS_RELEASE='-I/usr/include/python3.5m/'", 
-        "make -j3", "make install DESTDIR=/app"
+        "make",
```
Comment (TingPing) [inline net.ankiweb.Anki.json]: Well, `make` just defaults to a single job. Do `make -j \`nproc\`` to use available threads.

**Comment by TingPing** [comment]
bot, build net.ankiweb.Anki

**Comment by dsd** [comment]
Is the ARM build failure anything to worry about? It's not clear what it is failing on.

I'll try to upstream Makefile fixes to make it more suitable here. https://github.com/dae/anki/pull/227

File: bs4.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "bs4",
```
Comment (barthalion) [inline bs4.json]: Please prefix python libraries with `python-`. It's not a mistake per se but I have a script that looks for commonly used modules.

**Comment by barthalion** [comment]
bot, build net.ankiweb.Anki

**Comment by barthalion** [comment]
Also @dae, we will be more than happy to give you push access if you would like to have it.

File: bs4.json (hunk 4 above)
Comment (TingPing) [reply bs4.json]: These are all generated files for the python deps

**Comment by barthalion** [comment]
Build failure on ARM smells like race condition between installing file and creating directory where it should be installed. Try passing `-j1` to `make install`.

File: bs4.json (hunk 4 above)
Comment (barthalion) [reply bs4.json]: Too bad. I'll submit pull request for this later.

Comment (dsd) [reply bs4.json]: Yes, `flatpak-pip-generator` made those files for me.

**Comment by dsd** [comment]
The makefile is just a simple `install` recipe so I don't see that there is anything that could be racing.
https://github.com/dae/anki/blob/master/Makefile#L8

I can't see the github complaint about failing builds anymore. Does that mean that it succeeded upon rebuild?

**Comment by barthalion** [comment]
No, it's issue with builders not reporting back the results. I will try to reproduce the failure locally.

**Comment by dsd** [comment]
Ah, the failure is buried higher up in the log:
```
'designer/icons/media-record.png' -> '/app/share/anki/designer/icons/media-record.png'
cp: failed to preserve ownership for /app/share/anki/designer/icons/media-record.png: No such file or directory
```

but that's very weird indeed. Is the fs on the ARM builder corrupt or something?

File: net.ankiweb.Anki.json [hunk 5]
```diff
@@ -0,0 +1,117 @@
+{
+  "app-id": "net.ankiweb.Anki",
+  "runtime": "org.kde.Platform",
+  "runtime-version": "5.9",
```
Comment (barthalion) [inline net.ankiweb.Anki.json]: Btw, 5.10 is out.

Comment (dsd) [reply net.ankiweb.Anki.json]: Anki requires 5.9.x
https://github.com/dae/anki/blob/master/aqt/qt.py#L43

**Comment by barthalion** [comment]
> Is the fs on the ARM builder corrupt or something?
It's fishy that the error always happen on the same file though.

**Comment by dsd** [comment]
>  It's fishy that the error always happen on the same file though.

Have you confirmed this? I can only see a single build on https://flathub.org/builds/#/builders/5, not sure if we have seen this error multiple times

**Comment by dsd** [comment]
`media-record.png` is a symlink whereas the other files in that directory are regular files. If it is a consistent failure, then maybe that has something to do with it.

**Comment by barthalion** [comment]
https://flathub.org/builds/#/builders/78

File: bs4.json (hunk 4 above)
Comment (TingPing) [reply bs4.json]: https://github.com/flatpak/flatpak-builder-tools/commit/47340df0c5a8625ae6a20ab1db96bdaa2f323105

**Comment by dsd** [comment]
So what do you suggest as next steps here? Maybe one of the following?

* In the build manifest, delete the symlink before copying and then recreate it after using `ln -s`?
* Give me access to the ARM builder so I can login and investigate why it can't copy a symlink even though all the other builders are fine?
* Ship initially with no ARM build?

I think I've addressed all the review comments but please let me know if I'm missing anything else.

**Comment by nedrichards** [comment]
@dsd @ramcq can help you with the ARM builders.

**Comment by ramcq** [comment]
I've seen this error before in a flatpak build but hard to dig around for it from my phone. It was maybe something to do with trying to cp -a a symlink rather than just cp -r? I can have a check tomorrow.

**Comment by ramcq** [comment]
Yeah - https://github.com/flathub/org.freedesktop.Sdk.Extension.mono5/commit/e25f6b572ba5eab85c4eb1f1286b8301d3fd8253 or https://github.com/flathub/org.libreoffice.LibreOffice/commit/41b361bf4973b0981f834fe58b3f808b0ca81634 - maybe there's something weird about the kernel or FS on those ARM builders?

**Comment by barthalion** [comment]
@dsd Any chance you could look at fixing that `cp -a` invocation or disable `arm` for now?

**Comment by ramcq** [comment]
@barthalion There's no need to disable arm here, it's very easy to avoid this problem...

**Comment by barthalion** [comment]
Which I ask for now or later, when submitter can use `bot, build` himself.

**Comment by dsd** [comment]
pushed a change that should avoid the issue there

**Comment by nedrichards** [comment]
bot, build net.ankiweb.Anki

**Comment by nedrichards** [comment]
bot, build net.ankiweb.Anki

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/net.ankiweb.Anki

**Pull request closed without merging**